src/payload.rs

```diff
@@ -1,85 +1,76 @@
-use ring::digest;
-
+use crate::{crypto, DigestAlgorithm};
+use crate::error::*;
 /// A utility for hashing payloads. Feed your entity body to this, then pass the `finish`
 /// result to a request or response.
-pub struct PayloadHasher {
-    context: digest::Context,
-    algorithm: &'static digest::Algorithm,
-}
+pub struct PayloadHasher(Box<dyn crypto::Hasher>);
 
 impl PayloadHasher {
     /// Create a new PayloadHasher. The `content_type` should be lower-case and should
     /// not include parameters. The digest is assumed to be the same as the digest used
     /// for the credentials in the request.
-    pub fn new<B>(content_type: B, algorithm: &'static digest::Algorithm) -> Self
+    pub fn new<B>(content_type: B, algorithm: DigestAlgorithm) -> Result<Self>
```

Comment:
Another compatibility note.

Alternately, it looks like `CryptoError::UnsupportedDigest` is the only error that might be returned here, and that's arguably a programming error in this library or the custom `Cryptographer` implementation and thus suitable for a panic.  But I suppose at some point some Cryptographer implementation could conceivably want to signal an `Err`, so it probably makes sense to add `Result` now.

Reply:
Yeah, I don't think this is suitable for a panic TBH. It's reasonable to expect that all cryptographers don't support all algos here, and since there's no way to detect if the cryptographer supports a specific algo, yeah.